From 6f83e3fea99f8c90cc17dc04b6591910377a06e2 Mon Sep 17 00:00:00 2001
From: Sudheesh Mavila <sudheesh.mavila@amd.com>
Date: Tue, 12 Mar 2019 22:33:11 +0530
Subject: [PATCH 5763/5765] amd-xgbe Sometimes driver report incorrect link
 status and results in timeout error in driver Fix to EMBSWDEV-6825

Signed-off-by: Sudheesh Mavila <sudheesh.mavila@amd.com>
---
 drivers/net/ethernet/amd/xgbe/xgbe-phy-v2.c | 14 ++++++++++++--
 1 file changed, 12 insertions(+), 2 deletions(-)

diff --git a/drivers/net/ethernet/amd/xgbe/xgbe-phy-v2.c b/drivers/net/ethernet/amd/xgbe/xgbe-phy-v2.c
index 151bdb629e8a..ff4eeb55ab79 100755
--- a/drivers/net/ethernet/amd/xgbe/xgbe-phy-v2.c
+++ b/drivers/net/ethernet/amd/xgbe/xgbe-phy-v2.c
@@ -2565,8 +2565,18 @@ static int xgbe_phy_link_status(struct xgbe_prv_data *pdata, int *an_restart)
 	 */
 	reg = XMDIO_READ(pdata, MDIO_MMD_PCS, MDIO_STAT1);
 	reg = XMDIO_READ(pdata, MDIO_MMD_PCS, MDIO_STAT1);
-	if (reg & MDIO_STAT1_LSTATUS)
-		return 1;
+	if(phy_data->sfp_speed == XGBE_SFP_SPEED_10000) {
+		if ((reg & MDIO_STAT1_LSTATUS) && !(reg & MDIO_STAT1_FAULT)) {
+			return 1;
+		} else {
+			*an_restart = 1;
+			pdata->phy_if.phy_reset(pdata);
+			return 0;
+		}
+	} else {
+		if (reg & MDIO_STAT1_LSTATUS)
+			return 1;
+	}
 
 	/* No link, attempt a receiver reset cycle */
 	if (phy_data->rrc_count++ > XGBE_RRC_FREQUENCY) {
-- 
2.17.1

